Fixes issues #58, #59, #60

cgalvarez opened the pull request:
- Adds a description for linter options
- Adds the undocumented option `init`, which creates the `.codeclimate.yml` file if not found.
- Fixes #58
- Fixes #59 
- Fixes #60 

Please, read those issues to know about the problems the PR solves.

Commit:
adds desc for options, fixes AtomLinter/linter-codeclimate#60, fixes AtomLinter/linter-codeclimate#59, and fixes AtomLinter/linter-codeclimate#58

Arcanemagus commented on package.json:
```diff
@@ -11,7 +11,18 @@
+      "default": false,
+      "description": "Create a `.codeclimate.yml` file if not found."
+    },
+    "timeout": {
```
Change this to a boolean that just disables the timeout entirely, exactly like [this](https://github.com/AtomLinter/linter-flake8/blob/ab20e7c25a74f6d76607de2229834b97880e1aef/package.json#L26).

Arcanemagus commented on lib/index.js:
```diff
@@ -169,6 +168,7 @@ export default {
         const execArgs = ['-lc', cmd];
         const execOpts = {
           cwd: dirname(configurationFilePath),
+          timeout: 0,
```
You will need to observe the setting up in `activate()`, then conditionally add it [like this](https://github.com/AtomLinter/linter-flake8/blob/ab20e7c25a74f6d76607de2229834b97880e1aef/lib/main.js#L181-L183).

Arcanemagus commented on package.json:
```diff
@@ -11,7 +11,18 @@
     "executablePath": {
       "type": "string",
-      "default": "codeclimate"
+      "default": "/usr/local/bin/codeclimate",
```
How about instead we remove the existing check on the path, wrap the [call to exec](https://github.com/AtomLinter/linter-codeclimate/blob/d0c42f46717d19c2ed61b868eef0e676c3a4ebe5/lib/index.js#L173) in a `try`/`catch` block, and handle the `ENOENT` error if one is thrown?

It looks like this will potentially need to stop wrapping this in a call `/bin/bash`, I'm not sure of the implications of that though.

cgalvarez replied:
@Arcanemagus Do you mean something like:

```javascript
        let result;
        try {
          result = await Helpers.exec('/bin/bash', execArgs, execOpts);
        } catch (e) {
          if (e.code === 'ENOENT' && !badPaths.has(this.executablePath)) {
            const msg = 'codeclimate binary not found! Installation ' +
              'instructions at http://github.com/codeclimate/codeclimate';
            atom.notifications.addError(msg);
            // Only notify once per path
            badPaths.add(this.executablePath);
            return [];
          }
        }
```
instead [this](https://github.com/AtomLinter/linter-codeclimate/blob/d0c42f46717d19c2ed61b868eef0e676c3a4ebe5/lib/index.js#L173) and removing [this block](https://github.com/AtomLinter/linter-codeclimate/blob/d0c42f46717d19c2ed61b868eef0e676c3a4ebe5/lib/index.js#L78-L88)?

Got it working for both executablePath, `codeclimate` and `/usr/bin/local/codeclimate`.

Let me know what you think. I should also restore the default for `linter-codeclimate.executablePath` and change the description to `The path to the 'codeclimate' executable`.

Arcanemagus replied:
Exactly! Let the OS handle path resolution for us.

Although I notice you still have the executable as `/bin/bash`, is that necessary or can we just pass `codeclimate` in directly?

(I have yet to get this working on Windows as there is no wrapper script that I've found, so I can't test this locally)

Arcanemagus replied:
And restoring the default being part of that is why I put the comment on here, so it would be hidden when this gets fixed 😉.

Arcanemagus reviewed (changes requested):
Maybe @AtomLinter/linter-codeclimate can weigh in on why that attempt to read from `/dev/null` was in there?

maxjacobson commented on lib/index.js:
```diff
@@ -152,8 +152,7 @@ export default {
           '-f json',
           makeEngineString(linterEnabledEngines),
-          `'${atom.project.relativize(filePath)}'`,
-          '< /dev/null'].join(' ');
```
I'm not sure I understand why we had this `/dev/null` business previously, but I'm also not sure that it's hurting. It's been there all along, since: https://github.com/AtomLinter/linter-codeclimate/commit/1f78f75235029c22ad19eaa935d65f84830a6e1d

The CLI doesn't currently read anything from STDIN

cgalvarez replied:
@maxjacobson You're right, `< /dev/null` stuff didn't make it fail. I've just checked it locally. It all had to do with the timeout issue, but I changed both at the same time and that was my fault. Sorry. Reverting.

maxjacobson replied:
👍 sounds good. Thanks for looking into these improvements btw!

cgalvarez commented:
Now using codeclimate binary directly, instead of wrapping in a shell call. I've tested both, OS-managed (`codeclimate`) and full paths (like `/usr/local/bin/codeclimate`), in Ubuntu 16.04, and both works.

I've extracted the logic for checking if binary is not found to the function `ccBinaryFound()`, which is called on errors when creating the `.codeclimate.yml` file too.

I've also renamed and modified the function `makeEngineString()` to `getEngineArgs()` because of executing directly codeclimate requires its args to be in the form `['-f', 'json']` (providing them like `['-f json']` makes the execution return wrong output. Also extra empty args (like `""`, eventually returned by `makeEngineString()`) make the codeclimate execution to analyze the entire project folder.

Arcanemagus commented on lib/index.js:
```diff
@@ -6,9 +6,9 @@ import FS from 'fs';
+const getEngineArgs = engineNames => (
   engineNames.map(language => (`-e ${language}`))
-).join(' ');
+).join(' ').split(' ').filter(i => i);
```
Isn't this just the same as `.filter(i => i)`? Or does the join and split on the same string do something in the conversion?

cgalvarez replied:
Suppose `engineNames = ['engine1', 'engine2']`, then:

1. `map(...)` produces `['-e engine1', '-e engine2']`.
1. `join(...)` produces `'-e engine1 -e engine2'`.
1. `split(...)` produces `['-e', 'engine1', '-e', 'engine2']`.
1. `flter(...)` removes empty elements (if `engineNames = []`, then steps 1-3 produces `['']`, and this transforms it to `[]`).

If you remove `.join(' ').split(' ')` then you get result from step 1, except for the removal of the empty string from the resulting array (step 4), but executing directly codeclimate requires the args as in step 3.

Arcanemagus replied:
Ah, I guess this works just as well as a refactor to make the map push to a temporary array.

Arcanemagus commented on lib/index.js:
```diff
@@ -40,6 +40,18 @@ const endMeasure = (baseName) => {
+    badPaths.add(execPath);
+    return false;
+  }
+  return true;
```
This should probably re-`throw` errors it doesn't know how to handle.

Arcanemagus replied:
For example if the user mis-configures things and puts a directory in the `executablePath` setting, you get `EDIR` or `EACCESS` errors.

cgalvarez replied:
Why not to leverage the Atom notifications API in this case? Right now, since the shell call has been dropped, the error will be always related to the CodeClimate CLI. Instead my `ccBinaryFound()` function we could use next generic function:

```javascript
const notifyError = (cmd, err, msg, buttons = []) => {
  let friendlyDesc = '';
  switch (err.code) {
    case 'ENOENT':
      friendlyDesc = 'CodeClimate binary could not be found.'; break;
    case 'EACCES':
    case 'EDIR':
      friendlyDesc = 'Executable path not pointing to a binary.'; break;
    default: friendlyDesc = 'CodeClimate execution failed.';
  }
  if (!badCommands.has(cmd)) {
    atom.notifications.addError('linter-codeclimate error', {
      buttons: [{
        className: 'btn-install',
        onDidClick: () => {
          require('shell').openExternal('https://github.com/codeclimate/codeclimate');
        },
        text: 'Install guide',
      }].concat(buttons),
      detail: `${msg || friendlyDesc}
Exception details:
- CODE: ${err.code}
- MESSAGE: ${err.message}
- COMMAND: \`${cmd}\`
`,
      dismissable: true,
      stack: err.stack,
    });
    // Only notify once per path
    badCommands.add(cmd);
  }
};
```

which shows an error notification with the error code/message, the codeclimate command executed, an explanation and the stack trace, all without opening the dev console. Can be customized through the `switch()` and the optional parameters (message and buttons).

Changing `badPaths` to `badCommands` and using it to cache all the executed commands instead all the executable paths, the user can get a notification for each error only once, and in case of unhandled errors, you will always get the same info as before, but the generic message `CodeClimate execution failed`.

Here's how it looks:

![linter-codeclimate](https://cloud.githubusercontent.com/assets/10707639/25019435/d22daae4-208a-11e7-8c5c-23c83170b761.png)

Arcanemagus replied:
That looks amazing to me!

Arcanemagus commented on package.json:
```diff
@@ -11,7 +11,18 @@
       "type": "string",
-      "default": "codeclimate"
+      "default": "codeclimate",
+      "description": "The path to the `codeclimate` executable."
```
"The path" -> "The full path" seems to help users a little bit, I've yet to find a wording for this that is clear to everyone 😞.

cgalvarez replied:
Maybe "The absolute path"?

Arcanemagus replied:
The problem is that users quite often will stick a directory in there since using "path" to define the complete path of something is really only done by programmers from what I've seen.

Anyway, full or absolute works here 😉.

Arcanemagus commented on package.json:
```diff
@@ -11,7 +11,18 @@
+      "default": "codeclimate",
+      "description": "The path to the `codeclimate` executable."
+    },
+    "init": {
```
You might want to give this a `"title"`, otherwise Atom will just show it as "Init".

Arcanemagus reviewed (changes requested):
Getting close!

cgalvarez commented:
Is here any pending changes to approve?

Arcanemagus commented on lib/index.js:
```diff
@@ -149,28 +160,37 @@ export default {
+          if (ccBinaryFound(this.executablePath, e)) {
+            throw e;
+          }
+          return [];
```
Return `null` here, this may be a temporary issue so we should let `linter` keep managing the state of any messages that were currently there. If the code never had messages this doesn't change anything 😉.

Arcanemagus commented on lib/index.js:
```diff
@@ -149,28 +160,37 @@ export default {
-          '< /dev/null'].join(' ');
+        const execArgs = ['analyze', '-f', 'json'].concat(
+          getEngineArgs(linterEnabledEngines),
+          [atom.project.relativize(filePath), '</dev/null']);
```
Whoops, almost missed this. `project::relativize` isn't part of the public API, please switch this to using [`project::relativizePath`](https://atom.io/docs/api/v1.16.0/Project#instance-relativizePath). Note that you will need the first element from the array this returns, and you should handle the case where it returns `null` by falling back to the directory name of the current file.

cgalvarez replied:
I will need the 2nd element (last) from the array returned by `project::relativizePath`.

Let's take the filepath `/opt/cgalvarez/project/style.css` for example. `project::relativizePath` returns `["/opt/cgalvarez/project", "style.css"]`. We must feed the CodeClimate command with `"style.css"`, which is the relative path to the file to lint.

Also, the docs at the link you've provided doesn't say anything about returning `null`. While providing an absolute path as the input parameter, it will always return the array, and the `filePath` provided is the one [returned by the Atom textEditor for the currently open file](https://github.com/AtomLinter/linter-codeclimate/blob/master/lib/index.js#L90), so I think there is no need to handle that check :wink:.

Arcanemagus commented:
> Is here any pending changes to approve?

Apparently 😛 Sorry for the delay and the many reviews, as I said I can't test this locally so I'm having to re-review this fully each time trying to catch everything I can think of.

cgalvarez commented:
All done. Here you have a preview of the new function to notify errors (handled or not):
![captura de pantalla de 2017-04-15 14-55-09](https://cloud.githubusercontent.com/assets/10707639/25063700/0ad0c6f8-21ec-11e7-89d5-5f805233d101.png)

I've modified the function to show only the fields that exist or have been provided, and a docblock of rany doub could raise.

> Apparently :stuck_out_tongue: Sorry for the delay and the many reviews, as I said I can't test this locally so I'm having to re-review this fully each time trying to catch everything I can think of.

Don't worry. I forgot to make this PR in a new branch when I forked the repo (my fault!), that's why I need it to be merged before creating new branches the right way for two new PR I've prepared.

Arcanemagus commented on lib/index.js:
```diff
@@ -40,6 +39,55 @@ const endMeasure = (baseName) => {
 
+/**
+ * [notifyError description]
+ * @param  {Error}    err               The catched error.
```
"catched" -> "caught"

Arcanemagus commented on lib/index.js:
```diff
@@ -40,6 +39,55 @@ const endMeasure = (baseName) => {
 };
 
+/**
+ * [notifyError description]
```
"Show a clearer error in Atom when the exact problem is known."?

Arcanemagus commented on lib/index.js:
```diff
@@ -149,36 +195,42 @@ export default {
-            'Invalid JSON returned from codeclimate, error and output attached.',
-            e, result);
+          notifyError(e, `${this.executablePath} ${execArgs.join(' ')}`,
+            'Invalid JSON returned from CodeClimate.', `- RETURNED JSON: \`${result}\``);
```
Since the returned JSON can be quite large (as your example screenshot shows), let's just log it to the console here, and change the message to "See the Console for details".

cgalvarez commented:
@Arcanemagus I apologize for the lapse and the grammatical error :stuck_out_tongue_winking_eye:

Arcanemagus commented:
I don't see anything else wrong here just looking at it, any further changes that you can see needed @maxjacobson?

maxjacobson commented:
@Arcanemagus @cgalvarez thanks for working on this, including such a thorough review. The code LGTM, and if you've done local QA then I'm satisfied by that. I'm happy to do local QA myself if you'd like, but I'm actually not sure how to do that (it would be great to have those instructions in the README).

Arcanemagus committed:
Merge branch 'master' into master
